src/shared/tools.h

```diff
@@ -5,11 +5,7 @@
 #include <utility>
 #include <vector>
 #include <algorithm>
-
-#ifdef NULL
-#undef NULL
-#endif
-#define NULL 0
```

Comment:
What is the reason to replace a well defined `NULL` value with an implementation defined value? As we can read on https://en.cppreference.com/w/cpp/types/NULL the value of `NULL` can be `0` or `nullptr`. These two values have different sizes on 64 bit computers and not knowing which one it is can cause problems in some situations, like when using varargs where you specify which type you want to read but `NULL` makes sure that you don't know which type you give to the function.

I think a better change (if we want to change it) might be to define it as `nullptr` or `(void *)0` or `(size_t)0` to make sure that it has the same size as a pointer because `NULL` is often used there you would use a pointer. Another solution could be to ban `NULL` (but then we would have to change in a lot of places) by defining it to a value that will generate a compilation error when used. That could be done if we want to avoid confusion that people think they are using the standard library `NULL`  when they are really using our `NULL`.

Reply:
First of all, this partly solves the `-Wzero-as-null-pointer-constant` warning which is (yet?) not enabled in BN.

> NULL is often used there you would use a pointer.

Almost always, actually: indeed using `nullptr` would totally eliminate the issue of weird conversions.

> like when using varargs where you specify which type you want to read but NULL makes sure that you don't know which type you give to the function.

I fail to see how different sizes actually makes things act weirdly, but using `nullptr` (which _must_ be passable through varargs) would be nicer too.

> Another solution could be to ban NULL

I encourage everybody to use `nullptr`, but some files are still using pure C here: once a conversion of sorts is completed, then sure.


FWIW, while my `NULL` expands to `__null`, I know that MS's (and MINGW's, according to libcxx) expands to `0`: see this commit as a stop gap, because I would much rather see some platforms spit out warnings about NULL than giving an error.

Reply:
> I fail to see how different sizes actually makes things act weirdly

It's something I read somewhere but I'm not able to reproduce it so I don't know if it's actually a problem.

> but some files are still using pure C

If we make a new constant like `#define null ((void *)0)` that can be used in C too.

> my NULL expands to __null

And my `NULL` expands to `((void *)0)` on Debian/amd64/gcc :)

> I would much rather see some platforms spit out warnings about NULL than giving an error

Yes, it's probably a bad idea to hard ban `NULL` because there can be code (like in the standard library itself) that uses `NULL` that we don't easily control and then that code would break. For the same reason it might actually be a good idea to leave `NULL` defined as in the standard library because maybe there is some code in there that has expectations on the definition of `NULL`. I don't know.